tinygrad/codegen/uops.py

```diff
@@ -179,7 +182,84 @@ def fix_loop_scope(self, get_recursive_parents:Callable[..., Set[UOp]]):
     self.uops = flatten(loop_stack)
 
-  def uoptimize(self):
+  def replace_op(self, old, new):
+    for v in self.uops: v.vin = tuple(new if x is old else x for x in v.vin)
+    self.uops.remove(old)
+
+  def simplify_phi_loop(self, get_recursive_parents, loop_to_name, render_ops, ctx):
+    def alu_opposite(arg, x, y):
+      if arg == BinaryOps.ADD: return x - y
+      elif arg == BinaryOps.MUL: return Node.__floordiv__(x, y, False)
+      else: raise RuntimeError("unhandled alu")
+    def to_symbolic(u):
+      if u.uop == UOps.CONST: return NumNode(u.arg)
+      elif u.uop == UOps.LOOP: return Variable(loop_to_name[u], u.vin[0].arg, u.vin[1].arg)
+      elif u.uop == UOps.SPECIAL: return Variable(u.arg[1], 0, u.arg[2])
+      elif u.uop == UOps.ALU and u.arg == BinaryOps.ADD: return to_symbolic(u.vin[0]) + to_symbolic(u.vin[1])
+      elif u.uop == UOps.ALU and u.arg == BinaryOps.MUL: return to_symbolic(u.vin[0]) * to_symbolic(u.vin[1])
+      else: raise RuntimeError("unhandled op: " + u)
+    def loop_factor(get_recursive_parents, loop_to_name, with_loop: UOp, factored: Node, loop_op):
+      if with_loop == loop_op: return factored
+      elif with_loop.uop is UOps.ALU:
+        next_with_loop = next(v for v in with_loop.vin if v == loop_op or loop_op in get_recursive_parents(v))
+        non_loop = to_symbolic(next(v for v in with_loop.vin if v != next_with_loop and loop_op not in get_recursive_parents(v)))
+        return loop_factor(get_recursive_parents, loop_to_name, next_with_loop, alu_opposite(with_loop.arg, factored, non_loop), loop_op)
+    def const(x): return self.add(UOps.CONST, where.dtype, tuple(), x)
+    def mult_neg_1(x): return self.add(UOps.ALU, where.dtype, (x, const(-1)), BinaryOps.MUL)
+    def max(x, y): return self.add(UOps.ALU, where.dtype, (x, y), BinaryOps.MAX)
+
+    modified_phis = set()
+    for phi in [op for op in self.uops if op.uop is UOps.PHI]:
+      where_ops = [op for op in get_recursive_parents(phi) if op.arg == TernaryOps.WHERE and op.dtype is not None and dtypes.is_int(op.dtype)]
+      for where in sorted(where_ops, key=lambda x: self.uops.index(x)):
+        try:
+          loop_op = phi.vin[2]
+          comp, comp_lt, comp_gt = where.vin[0], where.vin[0].vin[0], where.vin[0].vin[1]
+          if where.vin[2].arg != 0 or comp.arg != BinaryOps.CMPLT or comp_gt.uop is not UOps.CONST or comp_gt.arg > 0: continue
+          factored = loop_factor(get_recursive_parents, loop_to_name, comp_lt, NumNode(comp_gt.arg), loop_op)
+        except (RuntimeError, StopIteration, IndexError, AssertionError): continue
+        final_value = NumNode(loop_op.vin[1].arg-1) - factored
+        self.uops, after_split_ops = self.uops[:(where_index:=self.uops.index(where))], self.uops[where_index:]
+        rendered = final_value.render(render_ops, ctx)
+        loop_length = loop_op.vin[1].arg - loop_op.vin[0].arg
+        clamped = mult_neg_1(max(const(-1*loop_length), mult_neg_1(max(rendered, const(0)))))
```

Comment:
Do you always need this max? Can't you check final_value.min and if it's >0 you can skip?